File: crates/kuben-api/src/routes/templates.rs
//! One-click templates (scenario 8): a small, reviewed catalogue of
//! single-process services. Generated credentials are stored in a Secret named
//! `<app>-credentials`; the App only *references* it, so no secret value ever
//! becomes part of an App spec, a release snapshot or the audit log.

use std::collections::BTreeMap;

use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use k8s_openapi::{ByteString, api::core::v1::Secret};
use kube::{
    Api,
    api::{DeleteParams, ObjectMeta, PostParams},
};
use kuben_core::{Error, perm::Perm};
use kuben_crd::{
    AppSpec, EnvVar, HealthCheck, KeyRef, Process, Protocol, Replicas, Runtime, Source, Volume, labels,
};
use rand::{Rng, distr::Alphanumeric};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::{apps, scope, validate};
use crate::{authz::Authz, error::ApiResult, state::ApiState};

/// Length of generated passwords and keys (alphanumeric, ~190 bits).
const SECRET_LEN: usize = 32;

enum Val {
    Lit(&'static str),
    /// A key of the credentials secret.
    Secret(&'static str),
}

struct Template {
    id: &'static str,
    name: &'static str,
    description: &'static str,
    category: &'static str,
    image: &'static str,
    port: u16,
    protocol: Protocol,
    command: &'static [&'static str],
    env: &'static [(&'static str, Val)],
    /// Secret keys generated as random strings.
    generated: &'static [&'static str],
    /// Secret keys derived from `{name}` and generated keys (`{password}`).
    derived: &'static [(&'static str, &'static str)],
    /// `(name, mount path, size)`.
    volumes: &'static [(&'static str, &'static str, &'static str)],
    health: Option<&'static str>,
    size: &'static str,
    /// For images that run as a non-root user and must write their volume.
    fs_group: Option<i64>,
}

use Val::{Lit, Secret as Sec};

const TEMPLATES: &[Template] = &[
    Template {
        id: "postgres",
        name: "PostgreSQL 17",
        description: "Relational database. Other apps connect with the `url` key of the credentials secret.",
        category: "database",
        image: "postgres:17-alpine",
        port: 5432,
        protocol: Protocol::Tcp,
        command: &[],
        env: &[
            ("POSTGRES_USER", Lit("app")),
            ("POSTGRES_DB", Lit("app")),
            ("POSTGRES_PASSWORD", Sec("password")),
            ("PGDATA", Lit("/var/lib/postgresql/data/pgdata")),
        ],
        generated: &["password"],
        derived: &[
            ("url", "postgres://app:{password}@{name}:5432/app"),
            ("host", "{name}"),
            ("port", "5432"),
            ("username", "app"),
            ("database", "app"),
        ],
        volumes: &[("data", "/var/lib/postgresql/data", "5Gi")],
        health: None,
        size: "small",
        fs_group: None,
    },
    Template {
        id: "redis",
        name: "Redis 7",
        description: "In-memory cache and queue with append-only persistence and a password.",
        category: "database",
        image: "redis:7-alpine",
        port: 6379,
        protocol: Protocol::Tcp,
        command: &[
            "sh",
            "-c",
            "exec redis-server --appendonly yes --requirepass \"$REDIS_PASSWORD\"",
        ],
        env: &[("REDIS_PASSWORD", Sec("password"))],
        generated: &["password"],
        derived: &[
            ("url", "redis://:{password}@{name}:6379/0"),
            ("host", "{name}"),
            ("port", "6379"),
        ],
        volumes: &[("data", "/data", "1Gi")],
        health: None,
        size: "nano",
        fs_group: None,
    },
    Template {
        id: "mariadb",
        name: "MariaDB 11",
        description: "MySQL-compatible database. Connect with the `url` key of the credentials secret.",
        category: "database",
        image: "mariadb:11",
        port: 3306,
        protocol: Protocol::Tcp,
        command: &[],
        env: &[
            ("MARIADB_DATABASE", Lit("app")),
            ("MARIADB_USER", Lit("app")),
            ("MARIADB_PASSWORD", Sec("password")),
            ("MARIADB_ROOT_PASSWORD", Sec("root-password")),
        ],
        generated: &["password", "root-password"],
        derived: &[
            ("url", "mysql://app:{password}@{name}:3306/app"),
            ("host", "{name}"),
            ("port", "3306"),
            ("username", "app"),
            ("database", "app"),
        ],
        volumes: &[("data", "/var/lib/mysql", "5Gi")],
        health: None,
        size: "medium",
        fs_group: None,
    },
    Template {
        id: "n8n",
        name: "n8n",
        description: "Workflow automation. Credentials are encrypted with a generated key.",
        category: "automation",
        image: "n8nio/n8n:stable",
        port: 5678,
        protocol: Protocol::Http,
        command: &[],
        env: &[
            ("N8N_ENCRYPTION_KEY", Sec("encryption-key")),
            ("N8N_PORT", Lit("5678")),
            ("GENERIC_TIMEZONE", Lit("UTC")),
        ],
        generated: &["encryption-key"],
        derived: &[],
        volumes: &[("data", "/home/node/.n8n", "1Gi")],
        health: Some("/healthz"),
        size: "small",
        fs_group: Some(1000),
    },
    Template {
        id: "uptime-kuma",
        name: "Uptime Kuma",
        description: "Self-hosted uptime monitoring with status pages.",
        category: "monitoring",
        image: "louislam/uptime-kuma:1",
        port: 3001,
        protocol: Protocol::Http,
        command: &[],
        env: &[],
        generated: &[],
        derived: &[],
        volumes: &[("data", "/app/data", "1Gi")],
        health: None,
        size: "small",
        fs_group: None,
    },
    Template {
        id: "vaultwarden",
        name: "Vaultwarden",
        description: "Bitwarden-compatible password manager. Sign-ups are closed; invite users from /admin with the generated admin token.",
        category: "security",
        image: "vaultwarden/server:latest",
        port: 80,
        protocol: Protocol::Http,
        command: &[],
        env: &[
            ("SIGNUPS_ALLOWED", Lit("false")),
            ("ADMIN_TOKEN", Sec("admin-token")),
        ],
        generated: &["admin-token"],
        derived: &[],
        volumes: &[("data", "/data", "1Gi")],
        health: Some("/alive"),
        size: "small",
        fs_group: None,
    },
    Template {
        id: "gitea",
        name: "Gitea",
        description: "Lightweight Git hosting (rootless image). Finish the installer on first visit.",
        category: "development",
        image: "gitea/gitea:1-rootless",
        port: 3000,
        protocol: Protocol::Http,
        command: &[],
        env: &[],
        generated: &[],
        derived: &[],
        volumes: &[
            ("data", "/var/lib/gitea", "5Gi"),
            ("config", "/etc/gitea", "100Mi"),
        ],
        health: Some("/api/healthz"),
        size: "small",
        fs_group: Some(1000),
    },
    Template {
        id: "whoami",
        name: "whoami",
        description: "Tiny HTTP echo service to test domains, TLS and routing.",
        category: "sample",
        image: "traefik/whoami:v1.10",
        port: 8080,
        protocol: Protocol::Http,
        command: &["/whoami", "--port", "8080"],
        env: &[],
        generated: &[],
        derived: &[],
        volumes: &[],
        health: None,
        size: "nano",
        fs_group: None,
    },
];

/// Name of the Secret holding a template app's generated credentials.
#[must_use]
pub fn credentials_secret(app: &str) -> String {
    format!("{app}-credentials")
}

fn random_secret() -> String {
    rand::rng()
        .sample_iter(&Alphanumeric)
        .take(SECRET_LEN)
        .map(char::from)
        .collect()
}

struct Rendered {
    spec: AppSpec,
    secret: BTreeMap<String, String>,
}

fn render(t: &Template, app: &str) -> Rendered {
    let mut secret: BTreeMap<String, String> = t
        .generated
        .iter()
        .map(|k| ((*k).to_owned(), random_secret()))
        .collect();
    for (key, pattern) in t.derived {
        let mut value = pattern.replace("{name}", app);
        for (k, v) in t
            .generated
            .iter()
            .filter_map(|k| secret.get(*k).map(|v| (*k, v.clone())))
        {
            value = value.replace(&format!("{{{k}}}"), &v);
        }
        secret.insert((*key).to_owned(), value);
    }
    let secret_name = credentials_secret(app);
    let env = t
        .env
        .iter()
        .map(|(name, value)| match value {
            Lit(v) => EnvVar {
                name: (*name).to_owned(),
                value: Some((*v).to_owned()),
                from_secret: None,
                from_service: None,
            },
            Sec(key) => EnvVar {
                name: (*name).to_owned(),
                value: None,
                from_secret: Some(KeyRef {
                    name: secret_name.clone(),
                    key: (*key).to_owned(),
                }),
                from_service: None,
            },
        })
        .collect();
    let spec = AppSpec {
        source: Source::from_image(t.image),
        runtime: Runtime {
            processes: BTreeMap::from([(
                "web".to_owned(),
                Process {
                    command: t.command.iter().map(ToString::to_string).collect(),
                    port: Some(t.port),
                    size: t.size.to_owned(),
                    replicas: Replicas::default(),
                    idle: None,
                    schedule: None,
                    time_zone: None,
                    protocol: t.protocol,
                },
            )]),
            health_check: t.health.map(|path| HealthCheck {
                path: path.to_owned(),
                port: None,
            }),
            fs_group: t.fs_group,
        },
        env,
        domains: Vec::new(),
        volumes: t
            .volumes
            .iter()
            .map(|(name, mount_path, size)| Volume {
                name: (*name).to_owned(),
                mount_path: (*mount_path).to_owned(),
                size: (*size).to_owned(),
                storage_class: None,
            })
            .collect(),
    };
    Rendered { spec, secret }
}

#[derive(Debug, Serialize, ToSchema)]
pub struct TemplateDto {
    pub id: String,
    pub name: String,
    pub description: String,
    pub category: String,
    pub image: String,
    pub port: u16,
    /// `http` (public route) or `tcp` (cluster-internal).
    pub protocol: String,
    /// `mountPath (size)` of each volume.
    pub volumes: Vec<String>,
    /// Keys of the `<app>-credentials` secret, e.g. `url`.
    pub connection_keys: Vec<String>,
}

fn template_dto(t: &Template) -> TemplateDto {
    let mut keys: Vec<String> = t.generated.iter().map(|k| (*k).to_owned()).collect();
    keys.extend(t.derived.iter().map(|(k, _)| (*k).to_owned()));
    keys.sort();
    TemplateDto {
        id: t.id.into(),
        name: t.name.into(),
        description: t.description.into(),
        category: t.category.into(),
        image: t.image.into(),
        port: t.port,
        protocol: if t.protocol.is_http() { "http" } else { "tcp" }.into(),
        volumes: t.volumes.iter().map(|(_, p, s)| format!("{p} ({s})")).collect(),
        connection_keys: keys,
    }
}

/// The template catalogue.
#[utoipa::path(
    get,
    path = "/templates", operation_id = "listTemplates",
    tag = "templates",
    responses((status = 200, body = Vec<TemplateDto>))
)]
pub async fn list(_authz: Authz) -> Json<Vec<TemplateDto>> {
    Json(TEMPLATES.iter().map(template_dto).collect())
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct DeployTemplate {
    /// App name (also the in-cluster hostname of TCP services).
    #[schema(example = "db")]
    pub name: String,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct DeployedTemplate {
    pub app: apps::AppDto,
    /// Secret with the generated credentials, e.g. `db-credentials`.
    pub credentials_secret: String,
    /// Reference it from other apps as `KEY=@<secret>/<key>`.
    pub connection_keys: Vec<String>,
}

/// Deploy a template into an environment.
#[utoipa::path(
    post,
    path = "/projects/{project}/environments/{environment}/templates/{template}", operation_id = "deployTemplate",
    tag = "templates",
    params(
        ("project" = String, Path, description = "Project name"),
        ("environment" = String, Path, description = "Environment short name"),
        ("template" = String, Path, description = "Template id"),
    ),
    request_body = DeployTemplate,
    responses(
        (status = 201, body = DeployedTemplate),
        (status = 404, body = crate::error::Problem),
        (status = 409, body = crate::error::Problem),
        (status = 422, body = crate::error::Problem),
    )
)]
pub async fn deploy(
    State(state): State<ApiState>,
    authz: Authz,
    Path((project, environment, template)): Path<(String, String, String)>,
    Json(body): Json<DeployTemplate>,
) -> ApiResult<(StatusCode, Json<DeployedTemplate>)> {
    let e = scope::environment(&state, &authz, &project, &environment)?;
    let _app = authz.require(&state, Perm::AppWrite, &e.chain())?;
    let _secret = authz.require(&state, Perm::SecretWrite, &e.chain())?;
    let t = TEMPLATES
        .iter()
        .find(|t| t.id == template)
        .ok_or_else(|| Error::NotFound(format!("template `{template}`")))?;
    validate::dns_label("name", &body.name, 40)?;
    let rendered = render(t, &body.name);
    apps::validate_spec(&rendered.spec)?;

    let secret_name = credentials_secret(&body.name);
    let secrets = Api::<Secret>::namespaced(scope::cluster(&state)?, &e.view.namespace);
    if secrets
        .get_opt(&secret_name)
        .await
        .map_err(|err| scope::kube_error(err, &secret_name))?
        .is_some()
    {
        return Err(Error::Conflict(format!("secret `{secret_name}` already exists")).into());
    }
    let object = Secret {
        metadata: ObjectMeta {
            name: Some(secret_name.clone()),
            namespace: Some(e.view.namespace.clone()),
            labels: Some(BTreeMap::from([
                (labels::MANAGED_BY.to_owned(), labels::MANAGER.to_owned()),
                (labels::ENVIRONMENT.to_owned(), e.view.name.clone()),
                (labels::APP.to_owned(), body.name.clone()),
            ])),
            ..ObjectMeta::default()
        },
        type_: Some("Opaque".into()),
        data: Some(
            rendered
                .secret
                .into_iter()
                .map(|(k, v)| (k, ByteString(v.into_bytes())))
                .collect(),
        ),
        ..Secret::default()
    };
    secrets
        .create(&PostParams::default(), &object)
        .await
        .map_err(|err| scope::kube_error(err, &secret_name))?;

    match apps::create_app(
        &state,
        &authz,
        &e,
        &body.name,
        rendered.spec,
        "template",
        Some(t.id.to_owned()),
    )
    .await
    {
        Ok(app) => Ok((
            StatusCode::CREATED,
            Json(DeployedTemplate {
                app,
                credentials_secret: secret_name,

File: internal/httpapi/templates.go
package api

// One-click templates (scenario 8, routes/templates.rs): a small, reviewed
// catalogue of single-process services. Generated credentials are stored in
// a Secret named `<app>-credentials`; the App only *references* it, so no
// secret value ever becomes part of an App spec, a release snapshot or the
// audit log.

import (
	"context"
	"crypto/rand"
	"fmt"
	"slices"
	"sort"
	"strings"

	corev1 "k8s.io/api/core/v1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/Teamtem-dev/kuben/api/v1alpha1"
	"github.com/Teamtem-dev/kuben/internal/core/kerrors"
	"github.com/Teamtem-dev/kuben/internal/core/opt"
	"github.com/Teamtem-dev/kuben/internal/core/perm"
	"github.com/Teamtem-dev/kuben/internal/httpapi/gen"
)

// SecretLen is the length of generated passwords and keys (alphanumeric, ~190 bits).
const SecretLen = 32

const alphanumericChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

// valKind tells a literal environment value from a generated secret.
type valKind int

const (
	valLit valKind = iota
	valSec
)

// envVal is a template environment value: literal text, or the key of a
// generated secret.
type envVal struct {
	Kind valKind
	Text string
}

func lit(s string) envVal { return envVal{Kind: valLit, Text: s} }
func sec(s string) envVal { return envVal{Kind: valSec, Text: s} }

// templateVolume is a volume a template mounts.
type templateVolume struct {
	Name      string
	MountPath string
	Size      string
}

// templateDerived is a secret value built from a pattern over the
// generated ones.
type templateDerived struct {
	Key     string
	Pattern string
}

// templateEnv is one environment variable of a template.
type templateEnv struct {
	Name string
	Val  envVal
}

// TemplateDef is a one-click service or database of the catalogue
// (routes/templates.rs).
type TemplateDef struct {
	ID          string
	Name        string
	Description string
	Category    string
	Image       string
	Port        uint16
	Protocol    v1alpha1.Protocol
	Command     []string
	Env         []templateEnv
	Generated   []string
	Derived     []templateDerived
	Volumes     []templateVolume
	// Health is the path of the HTTP health check, if the template has one.
	Health opt.Val[string]
	Size   string
	// FSGroup is set for images that run as a non-root user and must write
	// their volume.
	FSGroup opt.Val[int64]
}

// templatesCatalogue is the reviewed catalogue (TEMPLATES). It is built
// afresh on every call, so no caller can change what another one sees.
//
//nolint:funlen // one literal: the catalogue itself
func templatesCatalogue() []TemplateDef {
	return []TemplateDef{
		{
			ID:          "postgres",
			Name:        "PostgreSQL 17",
			Description: "Relational database. Other apps connect with the `url` key of the credentials secret.",
			Category:    "database",
			Image:       "postgres:17-alpine",
			Port:        5432,
			Protocol:    v1alpha1.ProtocolTCP,
			Command:     nil,
			Env: []templateEnv{
				{"POSTGRES_USER", lit("app")},
				{"POSTGRES_DB", lit("app")},
				{"POSTGRES_PASSWORD", sec("password")},
				{"PGDATA", lit("/var/lib/postgresql/data/pgdata")},
			},
			Generated: []string{"password"},
			Derived: []templateDerived{
				{"url", "postgres://app:{password}@{name}:5432/app"},
				{"host", "{name}"},
				{"port", "5432"},
				{"username", "app"},
				{"database", "app"},
			},
			Volumes: []templateVolume{{"data", "/var/lib/postgresql/data", "5Gi"}},
			Health:  opt.None[string](),
			Size:    "small",
			FSGroup: opt.None[int64](),
		},
		{
			ID:          "redis",
			Name:        "Redis 7",
			Description: "In-memory cache and queue with append-only persistence and a password.",
			Category:    "database",
			Image:       "redis:7-alpine",
			Port:        6379,
			Protocol:    v1alpha1.ProtocolTCP,
			Command: []string{
				"sh",
				"-c",
				"exec redis-server --appendonly yes --requirepass \"$REDIS_PASSWORD\"",
			},
			Env:       []templateEnv{{"REDIS_PASSWORD", sec("password")}},
			Generated: []string{"password"},
			Derived: []templateDerived{
				{"url", "redis://:{password}@{name}:6379/0"},
				{"host", "{name}"},
				{"port", "6379"},
			},
			Volumes: []templateVolume{{"data", "/data", "1Gi"}},
			Health:  opt.None[string](),
			Size:    "nano",
			FSGroup: opt.None[int64](),
		},
		{
			ID:          "mariadb",
			Name:        "MariaDB 11",
			Description: "MySQL-compatible database. Connect with the `url` key of the credentials secret.",
			Category:    "database",
			Image:       "mariadb:11",
			Port:        3306,
			Protocol:    v1alpha1.ProtocolTCP,
			Command:     nil,
			Env: []templateEnv{
				{"MARIADB_DATABASE", lit("app")},
				{"MARIADB_USER", lit("app")},
				{"MARIADB_PASSWORD", sec("password")},
				{"MARIADB_ROOT_PASSWORD", sec("root-password")},
			},
			Generated: []string{"password", "root-password"},
			Derived: []templateDerived{
				{"url", "mysql://app:{password}@{name}:3306/app"},
				{"host", "{name}"},
				{"port", "3306"},
				{"username", "app"},
				{"database", "app"},
			},
			Volumes: []templateVolume{{"data", "/var/lib/mysql", "5Gi"}},
			Health:  opt.None[string](),
			Size:    "medium",
			FSGroup: opt.None[int64](),
		},
		{
			ID:          "n8n",
			Name:        "n8n",
			Description: "Workflow automation. Credentials are encrypted with a generated key.",
			Category:    "automation",
			Image:       "n8nio/n8n:stable",
			Port:        5678,
			Protocol:    v1alpha1.ProtocolHTTP,
			Command:     nil,
			Env: []templateEnv{
				{"N8N_ENCRYPTION_KEY", sec("encryption-key")},
				{"N8N_PORT", lit("5678")},
				{"GENERIC_TIMEZONE", lit("UTC")},
			},
			Generated: []string{"encryption-key"},
			Derived:   nil,
			Volumes:   []templateVolume{{"data", "/home/node/.n8n", "1Gi"}},
			Health:    opt.Some("/healthz"),
			Size:      "small",
			FSGroup:   opt.Some[int64](1000),
		},
		{
			ID:          "uptime-kuma",
			Name:        "Uptime Kuma",
			Description: "Self-hosted uptime monitoring with status pages.",
			Category:    "monitoring",
			Image:       "louislam/uptime-kuma:1",
			Port:        3001,
			Protocol:    v1alpha1.ProtocolHTTP,
			Command:     nil,
			Env:         nil,
			Generated:   nil,
			Derived:     nil,
			Volumes:     []templateVolume{{"data", "/app/data", "1Gi"}},
			Health:      opt.None[string](),
			Size:        "small",
			FSGroup:     opt.None[int64](),
		},
		{
			ID:          "vaultwarden",
			Name:        "Vaultwarden",
			Description: "Bitwarden-compatible password manager. Sign-ups are closed; invite users from /admin with the generated admin token.",
			Category:    "security",
			Image:       "vaultwarden/server:latest",
			Port:        80,
			Protocol:    v1alpha1.ProtocolHTTP,
			Command:     nil,
			Env: []templateEnv{
				{"SIGNUPS_ALLOWED", lit("false")},
				{"ADMIN_TOKEN", sec("admin-token")},
			},
			Generated: []string{"admin-token"},
			Derived:   nil,
			Volumes:   []templateVolume{{"data", "/data", "1Gi"}},
			Health:    opt.Some("/alive"),
			Size:      "small",
			FSGroup:   opt.None[int64](),
		},
		{
			ID:          "gitea",
			Name:        "Gitea",
			Description: "Lightweight Git hosting (rootless image). Finish the installer on first visit.",
			Category:    "development",
			Image:       "gitea/gitea:1-rootless",
			Port:        3000,
			Protocol:    v1alpha1.ProtocolHTTP,
			Command:     nil,
			Env:         nil,
			Generated:   nil,
			Derived:     nil,
			Volumes: []templateVolume{
				{"data", "/var/lib/gitea", "5Gi"},
				{"config", "/etc/gitea", "100Mi"},
			},
			Health:  opt.Some("/api/healthz"),
			Size:    "small",
			FSGroup: opt.Some[int64](1000),
		},
		{
			ID:          "whoami",
			Name:        "whoami",
			Description: "Tiny HTTP echo service to test domains, TLS and routing.",
			Category:    "sample",
			Image:       "traefik/whoami:v1.10",
			Port:        8080,
			Protocol:    v1alpha1.ProtocolHTTP,
			Command:     []string{"/whoami", "--port", "8080"},
			Env:         nil,
			Generated:   nil,
			Derived:     nil,
			Volumes:     nil,
			Health:      opt.None[string](),
			Size:        "nano",
			FSGroup:     opt.None[int64](),
		},
	}
}

// CredentialsSecret is the name of the Secret holding a template app's generated credentials.
func CredentialsSecret(app string) string {
	return fmt.Sprintf("%s-credentials", app)
}

func randomSecret() (string, error) {
	const maxByte = 256 - (256 % len(alphanumericChars))
	b := make([]byte, 1)
	out := make([]byte, SecretLen)
	for i := 0; i < SecretLen; {
		if _, err := rand.Read(b); err != nil {
			return "", fmt.Errorf("generate a credential: %w", err)
		}
		if int(b[0]) < maxByte {
			out[i] = alphanumericChars[int(b[0])%len(alphanumericChars)]
			i++
		}
	}
	return string(out), nil
}

// RenderedTemplate is a template made into an app: its spec and the values
// of the Secret it reads.
type RenderedTemplate struct {
	Spec   v1alpha1.AppSpec
	Secret map[string]string
}

func renderTemplate(t TemplateDef, app string) (RenderedTemplate, error) {
	secret := make(map[string]string, len(t.Generated)+len(t.Derived))
	for _, k := range t.Generated {
		val, err := randomSecret()
		if err != nil {
			return RenderedTemplate{}, err
		}
		secret[k] = val
	}
	for _, d := range t.Derived {
		val := strings.ReplaceAll(d.Pattern, "{name}", app)
		for _, k := range t.Generated {
			if secVal, ok := secret[k]; ok {
				val = strings.ReplaceAll(val, "{"+k+"}", secVal)
			}
		}
		secret[d.Key] = val
	}

	secretName := CredentialsSecret(app)
	envVars := make([]v1alpha1.EnvVar, len(t.Env))
	for i, e := range t.Env {
		switch e.Val.Kind {
		case valLit:
			v := e.Val.Text
			envVars[i] = v1alpha1.EnvVar{Name: e.Name, Value: &v}
		case valSec:
			envVars[i] = v1alpha1.EnvVar{
				Name: e.Name,
				FromSecret: &v1alpha1.KeyRef{
					Name: secretName,
					Key:  e.Val.Text,
				},
			}
		}
	}

	volumes := make([]v1alpha1.Volume, len(t.Volumes))
	for i, v := range t.Volumes {
		volumes[i] = v1alpha1.Volume{
			Name:      v.Name,
			MountPath: v.MountPath,
			Size:      v.Size,
		}
	}

	var healthCheck *v1alpha1.HealthCheck
	if path, ok := t.Health.Get(); ok {
		healthCheck = &v1alpha1.HealthCheck{Path: path}
	}

	port := t.Port
	spec := v1alpha1.AppSpec{
		Source: v1alpha1.SourceFromImage(t.Image),
		Runtime: v1alpha1.Runtime{
			Processes: map[string]v1alpha1.Process{
				"web": {
					Command:  slices.Clone(t.Command),
					Port:     &port,
					Size:     t.Size,
					Replicas: v1alpha1.DefaultReplicas(),
					Protocol: t.Protocol,
				},
			},
			HealthCheck: healthCheck,
			FSGroup:     t.FSGroup.Ptr(),
		},
		Env:     envVars,
		Volumes: volumes,
	}

	return RenderedTemplate{Spec: spec, Secret: secret}, nil
}

func templateDto(t TemplateDef) gen.TemplateDto {
	keys := make([]string, 0, len(t.Generated)+len(t.Derived))
	keys = append(keys, t.Generated...)
	for _, d := range t.Derived {
		keys = append(keys, d.Key)
	}
	sort.Strings(keys)

	volumes := make([]string, len(t.Volumes))
	for i, v := range t.Volumes {
		volumes[i] = fmt.Sprintf("%s (%s)", v.MountPath, v.Size)
	}

	protocol := "http"
	if !t.Protocol.IsHTTP() {
		protocol = "tcp"
	}

	return gen.TemplateDto{
		Category:       t.Category,
		ConnectionKeys: keys,
		Description:    t.Description,
		ID:             t.ID,
		Image:          t.Image,
		Name:           t.Name,
		Port:           int32(t.Port),
		Protocol:       protocol,
		Volumes:        volumes,
	}
}

func findTemplate(id string) (TemplateDef, bool) {
	catalogue := templatesCatalogue()
	if i := slices.IndexFunc(catalogue, func(t TemplateDef) bool { return t.ID == id }); i >= 0 {
		return catalogue[i], true
	}
	return TemplateDef{}, false
}

// ListTemplates returns the template catalogue.
func (s *Server) ListTemplates(ctx context.Context) ([]gen.TemplateDto, error) {
	if _, err := s.access(ctx); err != nil {
		return nil, err
	}
	catalogue := templatesCatalogue()
	dtos := make([]gen.TemplateDto, len(catalogue))
	for i, t := range catalogue {
		dtos[i] = templateDto(t)
	}
	return dtos, nil
}

// DeployTemplate deploys a template into an environment.
func (s *Server) DeployTemplate(
	ctx context.Context, req *gen.DeployTemplate, params gen.DeployTemplateParams,
) (gen.DeployTemplateRes, error) {
	if req == nil {
		return nil, kerrors.New(kerrors.Validation, "missing request body")
	}
	acc, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	e, err := s.findEnvironment(ctx, acc, params.Project, params.Environment)
	if err != nil {
		return nil, err
	}
	chain := e.chain()
	if _, err := acc.Require(perm.AppWrite, chain); err != nil {
		return nil, err //nolint:wrapcheck // a kerrors already
	}
	if _, err := acc.Require(perm.SecretWrite, chain); err != nil {
		return nil, err //nolint:wrapcheck // a kerrors already
	}
	t, found := findTemplate(params.Template)
	if !found {
		return nil, scopeNotFound("template", params.Template)
	}
	if err := DNSLabel("name", req.Name, 40); err != nil {
		return nil, err
	}
	rendered, err := renderTemplate(t, req.Name)
	if err != nil {
		return nil, err
	}
	if err := validateSpec(&rendered.Spec); err != nil {
		return nil, err
	}

	secretName := CredentialsSecret(req.Name)
	secrets, err := s.createCredentials(ctx, e, req.Name, rendered.Secret)
	if err != nil {
		return nil, err
	}

	appDto, err := s.createApp(ctx, acc, e, req.Name, rendered.Spec)
	if err != nil {
		s.dropOrphanedCredentials(ctx, secrets, secretName)
		return nil, err
	}

	return &gen.DeployedTemplate{
		App:               appDto,
		CredentialsSecret: secretName,
		ConnectionKeys:    templateDto(t).ConnectionKeys,
	}, nil
}

// createCredentials stores the generated values of app in its credentials
// Secret, which must not exist yet, and returns the client to remove it with.
func (s *Server) createCredentials(
	ctx context.Context, e envScope, app string, values map[string]string,
) (secretDeleter, error) {
	cluster, err := s.cluster()
	if err != nil {
		return nil, err
	}
	secretName := CredentialsSecret(app)
	namespace := e.namespace()
	secrets := cluster.Typed.CoreV1().Secrets(namespace)

	_, err = secrets.Get(ctx, secretName, metav1.GetOptions{})
	if err == nil {
		return nil, kerrors.New(kerrors.Conflict, "secret `%s` already exists", secretName)
	} else if !apierrors.IsNotFound(err) {
		return nil, kubeError(err, secretName)
	}

	data := make(map[string][]byte, len(values))
	for k, v := range values {
		data[k] = []byte(v)
	}
	object := &corev1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name:      secretName,
			Namespace: namespace,
			Labels: map[string]string{
				v1alpha1.LabelManagedBy:   v1alpha1.LabelManagerValue,
				v1alpha1.LabelEnvironment: e.resourceName(),
				v1alpha1.LabelApp:         app,
			},
		},
		Type: corev1.SecretTypeOpaque,
		Data: data,
	}
	if _, err := secrets.Create(ctx, object, metav1.CreateOptions{}); err != nil {
		return nil, kubeError(err, secretName)
	}
	return secrets, nil
}

// secretDeleter is the part of the Secrets client the cleanup needs.
type secretDeleter interface {
	Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
}

// dropOrphanedCredentials removes the credentials of an app that could not
// be created, so none are left behind. It runs even when the request was
// cancelled; a failure is logged and does not change the answer, which stays
// the creation's error (Rust discarded the failure).
func (s *Server) dropOrphanedCredentials(ctx context.Context, secrets secretDeleter, name string) {
	err := secrets.Delete(context.WithoutCancel(ctx), name, metav1.DeleteOptions{})
	if err != nil && !apierrors.IsNotFound(err) {
		s.deps.Logger.Error("orphaned template credentials were not removed", "secret", name, "error", err)
	}
}
